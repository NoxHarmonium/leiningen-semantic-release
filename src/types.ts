import { either } from "fp-ts/lib/Either";
import { lstatSync } from "fs";
import * as t from "io-ts";
import { withFallback } from "io-ts-types/lib/withFallback";
import { Context } from "semantic-release";

const directoryPathCodec = new t.Type<string, string, unknown>(
  "DirectoryPath",
  t.string.is,
  (u, c) =>
    either.chain(t.string.validate(u, c), s => {
      return lstatSync(s).isDirectory()
        ? t.failure(u, c, "not a path to a valid directory")
        : t.success(s);
    }),
  String
);

export type DirectoryPath = t.TypeOf<typeof directoryPathCodec>;

export const pluginConfigCodec = t.type({
  skipDeploy: withFallback(t.boolean, true),
  pkgRoot: directoryPathCodec,
  uberJar: withFallback(t.boolean, false)
});

export type PluginConfig = t.TypeOf<typeof pluginConfigCodec>;

export type PluginStep = (
  config: PluginConfig,
  context: Context
) => Promise<void>;
