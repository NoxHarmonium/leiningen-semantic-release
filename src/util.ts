import execa from "execa";
import { Context } from "semantic-release";
import * as pkg from "../package.json";

/**
 * Execute while streaming to stdout in realtime
 */
export const exec = async (
  file: string,
  args: readonly string[],
  cwd: string
) => {
  const childProcess = execa(file, args, { cwd });
  if (childProcess.stdout !== null) {
    childProcess.stdout.pipe(process.stdout);
  }
  if (childProcess.stderr !== null) {
    childProcess.stderr.pipe(process.stderr);
  }
  return childProcess.then(({ stdout }) => stdout);
};

export const printVersion = ({ logger }: Context) => {
  logger.log(`Running ${pkg.name} version ${pkg.version}`);
};
