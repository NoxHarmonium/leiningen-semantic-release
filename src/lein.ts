import { Context } from "semantic-release";
import { PluginConfig } from "./types";
import { exec } from "./util";

/**
 * Gets the path to the "lein" executable.
 * Could be extended in the future to try different locations.
 */
const findCommand = async () => "lein"

export const updateVersionInLeinProject = async (
  { pkgRoot }: PluginConfig,
  { logger, nextRelease }: Context
) => {
  if (nextRelease === undefined) {
    throw new Error("context.nextRelease is undefined");
  }
  const version = nextRelease.version;
  logger.log(`Updating project.clj to version ${version}`);
  const command = await findCommand();
  return exec(command, ["change", "version", "set", `"${version}"`], pkgRoot);
};

export const deploy = async (
  { pkgRoot }: PluginConfig,

  { logger, nextRelease }: Context
) => {
  if (nextRelease === undefined) {
    throw new Error("context.nextRelease is undefined");
  }
  const version = nextRelease.version;
  logger.log("Deploying version %s with leiningen", version);
  const command = await findCommand();
  return exec(command, ["deploy"], pkgRoot);
};

export const getVersion = async (
  { pkgRoot }: PluginConfig,
  { logger }: Context
) => {
  logger.log("Sanity checking project version with leiningen");
  const command = await findCommand();
  return exec(
    command,
    // Uses the lein-project-version plugin to print the version number
    // Only updates the project in memory, the project file on disk will remain unaffected
    [
      "update-in",
      ":plugins",
      "conj",
      '[lein-project-version "0.1.0"]',
      "--",
      "project-version"
    ],
    pkgRoot
  );
};

export const leinPackage = async (
  { uberJar, pkgRoot }: PluginConfig,
  { logger }: Context
) => {
  logger.log("Packaging with leiningen");
  const command = await findCommand();
  return exec(command, [uberJar ? "uberjar" : "jar"], pkgRoot);
};
