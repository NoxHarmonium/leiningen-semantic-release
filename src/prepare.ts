import { Context } from "semantic-release";
import { leinPackage, updateVersionInLeinProject } from "./lein";
import { PluginConfig, PluginStep } from "./types";
import { printVersion } from "./util";

export const prepareStep: PluginStep = async (
  config: PluginConfig,
  context: Context
) => {
  printVersion(context);

  // set and commit version number in pom.xml
  await updateVersionInLeinProject(config, context);
  await leinPackage(config, context);
};
