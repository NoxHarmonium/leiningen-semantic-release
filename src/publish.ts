import { Context } from "semantic-release";
import { deploy } from "./lein";
import { PluginConfig, PluginStep } from "./types";
import { printVersion } from "./util";

export const publishStep: PluginStep = async (
  config: PluginConfig,
  context: Context
) => {
  printVersion(context);

  if (!config.skipDeploy) {
    // deploy the project to maven-central
    await deploy(config, context);
  }
};
