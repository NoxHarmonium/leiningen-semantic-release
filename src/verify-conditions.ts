import { Context } from "semantic-release";
import { getVersion } from "./lein";
import { PluginConfig, PluginStep } from "./types";
import { printVersion } from "./util";

/**
 * Verify that the leiningen project is properly setup to allow deployment
 */
export const verifyConditionsStep: PluginStep = async (
  config: PluginConfig,
  context: Context
) => {
  printVersion(context);

  await getVersion(config, context);
};
