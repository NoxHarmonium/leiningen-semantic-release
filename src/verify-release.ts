import { Context } from "semantic-release";
import { valid } from "semver";
import { getVersion } from "./lein";
import { PluginConfig, PluginStep } from "./types";
import { printVersion } from "./util";

export const verifyReleaseStep: PluginStep = async (
  config: PluginConfig,
  context: Context
) => {
  const { logger } = context;
  printVersion(context);

  const projectVersion = await getVersion(config, context);

  // check integrity of pom version
  if (valid(projectVersion) === null) {
    logger.log(
      "WARNING: project.clj version of %s is an invalid semver version",
      projectVersion
    );
  }
};
