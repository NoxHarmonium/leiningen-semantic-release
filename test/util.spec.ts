import { writeFileSync } from "fs";
import { basename, dirname } from "path";
import * as tempy from "tempy";
import { inspect } from "util";

jest.mock("../package.json", () => ({
  name: "test name",
  version: "test version"
}));

import { exec, printVersion } from "../src/util";

// tslint:disable: no-identical-functions

describe("when executing an external process", () => {
  describe("when running a successful command", () => {
    it("should return stdout as a trimmed string", async () => {
      const result = await exec(
        "node",
        ["-e", "console.log('hello from unit testing!')"],
        "."
      );
      expect(result).toMatchSnapshot();
    });
  });
  describe("when an error occurs", () => {
    it("should catch the error", async () => {
      expect.assertions(1);
      try {
        await exec(
          "node",
          ["-e", "throw new Error('synthetic unit test error')"],
          "."
        );
      } catch (e) {
        expect(e).toBeInstanceOf(Error);
      }
    });
  });
  describe("when using a different working directory", () => {
    it("should should execute command relative to that directory", async () => {
      const currentDirTest = await exec(
        "node",
        [
          "-e",
          `console.dir(require('fs').lstatSync(${inspect(
            __filename
          )}).isFile().toString())`
        ],
        __dirname
      );
      const relativePathTest = await exec(
        "node",
        ["-e", "console.dir(require('fs').lstatSync('config.yml').isFile())"],
        "src/../test/../.circleci"
      );
      const tempPath = tempy.file();

      writeFileSync(tempPath, "test");
      const tempFile = basename(tempPath);
      const tempDir = dirname(tempPath);
      const randomTest = await exec(
        "node",
        [
          "-e",
          `console.dir(require('fs').lstatSync(${inspect(tempFile)}).isFile())`
        ],
        tempDir
      );

      expect(currentDirTest).toMatchSnapshot();
      expect(relativePathTest).toMatchSnapshot();
      expect(randomTest).toMatchSnapshot();
    });
  });
  describe("when printing version", () => {
    it("should print the version and name stored in package.json (mocked)", async () => {
      expect.assertions(1);
      printVersion({
        logger: {
          log: (message: string, _: readonly unknown[]) => {
            expect(message).toMatchSnapshot();
          },
          error: (_: string, __: readonly unknown[]) => {
            throw new Error("Should not be called");
          }
        }
      });
    });
  });
});
