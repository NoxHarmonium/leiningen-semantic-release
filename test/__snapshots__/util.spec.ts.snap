// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when executing an external process when printing version should print the version and name stored in package.json (mocked) 1`] = `"Running test name version test version"`;

exports[`when executing an external process when running a successful command should return stdout as a trimmed string 1`] = `"hello from unit testing!"`;

exports[`when executing an external process when using a different working directory should should execute command relative to that directory 1`] = `"'true'"`;

exports[`when executing an external process when using a different working directory should should execute command relative to that directory 2`] = `"true"`;

exports[`when executing an external process when using a different working directory should should execute command relative to that directory 3`] = `"true"`;
